### DISCUSSION TITLE
Dependencies and versioning

### DISCUSSION THREAD

**Discussion by soundstep** [teambit/bit#4467] [Help] [answered]
Hi,
A question about dependencies, as Bit does its versioning per component.
In the case of cross-component dependencies where you have a basic component like a button used in many other components, I can see it quickly turning into a dependency management chaos where one component may be using a newer version of the button than another component.
Unless I am missing something? Is there a specific approach to release all of them as a whole and still benefit from their usage separated by component versions?

**Comment by itaymendel** [comment, marked as answer]
> For any reader looking to understand the reasoning behind per-component versioning, [see here](https://medium.com/eightshapes-llc/versioning-design-systems-48cceb5ace4d)

While dependencies are per-component, Bit sets versions according to what is resolved in your workspace' (ie - what's in `node_modules`), according to the resolution policy configured. Now, what happens in your workspace is essentially a mono-repo like setup, but with a more granular level of components. All components have generated modules (open `node_modules`, see that you have the components' dists there). So when dependencies resolved, their versions are resolved from what's in the workspace.  
As a result, all components depending on button will get the same version of button.

This also means that when you modify a component you can see how it effects on the rest of the components. If you modify a component that has dependents, you can run `bit status`, and see how its modification will affect other components and their future versions (think about a "splash zone" where when I change something Bit shows what "gets wet"). This process is done recursively.

If you want to set one or more of your components to depend on a specific version of another, you can do that. You need to add a `variant` in your workspace for these components, and there set a different dependency resolution policy.

**Reply by soundstep** [reply]
@itaymendel I see, very interesting, is that one of the reasons why you went to this design as opposed to having the components package name in the `package.json` with semver versions?

**Reply by itaymendel** [reply]
I'm not sure I completely follow your question, but I'll try to answer to the best of my ability.
Bit generates a `package.json` for each component (otherwise it will not be able to support an `npm i <component>` nor using components as node modules). However, the design approach we took was that you manage a `workspace.jsonc` file that applies policies of configurations according to component names / locations in the workspace.
Moreover, dependencies are fetched from the source code of each components with static code analysis.

Regarding semver versions in `package.json`, I assume you mean the `package.json`s in `node_modules` dir? (per component). we don't keep a number there as Bit calculates that number according to the version of that component in the workspace (behind the scenes Bit keeps two lists of dependencies per component; packages and other components, and there are some diffs in how they are handled).

**Reply by soundstep** [reply]
What I meant was that the components are not part of the "package.json", as dependencies like:
```
"dependencies": {
  "@bit/my-components.button": "^0.8.0"
```
But I see what's happening, thanks for answering.
